src/systems/particle_emitter/ParticleEmitter.cc

```diff
@@ -0,0 +1,335 @@
+  public: ignition::transport::Node node;
+
+  /// \brief Particle emitter entity.
+  public: Entity emitterEntity{kNullEntity};
+
+  /// \brief The particle emitter command requested externally.
+  public: ignition::msgs::ParticleEmitter userCmd;
+
+  public: bool newDataReceived = false;
+
+  /// \brief A mutex to protect the user command.
+  public: std::mutex mutex;
+};
+
+//////////////////////////////////////////////////
+void ParticleEmitterPrivate::OnCmd(const msgs::ParticleEmitter &_msg)
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
+  this->userCmd = _msg;
+  this->newDataReceived = true;
+}
+
+//////////////////////////////////////////////////
+ParticleEmitter::ParticleEmitter()
+  : System(), dataPtr(std::make_unique<ParticleEmitterPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter::Configure(const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &_eventMgr)
+{
+  Model model = Model(_entity);
+
+  if (!model.Valid(_ecm))
+  {
+    ignerr << "ParticleEmitter plugin should be attached to a model entity. "
+           << "Failed to initialize." << std::endl;
+    return;
+  }
+
+  // Create a particle emitter entity.
+  this->dataPtr->emitterEntity = _ecm.CreateEntity();
+  if (this->dataPtr->emitterEntity == kNullEntity)
+  {
+    ignerr << "Failed to create a particle emitter entity" << std::endl;
+    return;
+  }
+
+  // allow_renaming
+  bool allowRenaming = false;
+  if (_sdf->HasElement("allow_renaming"))
+    allowRenaming = _sdf->Get<bool>("allow_renaming");
+
+  // Name.
+  std::string name = "particle_emitter_entity_" +
+      std::to_string(this->dataPtr->emitterEntity);
+  if (_sdf->HasElement("emitter_name"))
+  {
+    std::set<std::string> emitterNames;
+    std::string emitterName = _sdf->Get<std::string>("emitter_name");
+
+    // check to see if name is already taken
+    _ecm.Each<components::Name, components::ParticleEmitter>(
+        [&emitterNames](const Entity &, const components::Name *_name,
+                      const components::ParticleEmitter *)
+        {
+          emitterNames.insert(_name->Data());
+          return true;
+        });
+
+    name = emitterName;
+
+    // rename emitter if needed
+    if (emitterNames.find(emitterName) != emitterNames.end())
+    {
+      if (!allowRenaming)
+      {
+        ignwarn << "Entity named [" << name
+                << "] already exists and "
+                << "[allow_renaming] is false. Entity not spawned."
+                << std::endl;
+        return;
+      }
+      int counter = 0;
+      while (emitterNames.find(name) != emitterNames.end())
+      {
+        name = emitterName + "_" + std::to_string(++counter);
+      }
+      ignmsg << "Entity named [" << emitterName
+             << "] already exists. Renaming it to " << name << std::endl;
+    }
+  }
+  this->dataPtr->emitter.set_name(name);
```

Comment:
We should probably leave a note that the `name` field in the emitter message never seems to be used past this point (at least, I don't see the message's `name` being used anywhere else in this PR).

Reply:
The name is also used in the `component` and topic name. I'll leave this code as-is.